src/gpu/nvidia/cudnn_matmul_executor.hpp

```diff
@@ -334,43 +372,67 @@ struct cudnn_matmul_bias_exec_t : public cudnn_matmul_exec_base_t {
                     = CTX_IN_SYCL_MEMORY(DNNL_ARG_ATTR_SCALES | DNNL_ARG_DST);
 
             interop_task(matmul_impl_, engine, cgh, cuda_stream, arg_wt,
-                    arg_src, arg_dst, arg_bias, /*nullptr*/ arg_scratch,
+                    arg_src, arg_dst, arg_bias, arg_algo_scratch,
+                    arg_bias_scratch, arg_block_a_scratch, arg_block_b_scratch,
+                    arg_block_c_scratch, scaled_arg_src, scaled_arg_wt,
                     arg_src_scale, arg_wei_scale, arg_dst_scale);
         });
     }
-};
-
-struct cudnn_matmul_exec_t : public cudnn_matmul_exec_base_t {
-    status_t execute(const exec_ctx_t &ctx, impl::engine_t *engine,
-            const std::shared_ptr<cudnn_matmul_impl_t> matmul_impl_,
-            std::size_t scratchpad_size) override {
-
-        nvidia::stream_t *cuda_stream
-                = utils::downcast<nvidia::stream_t *>(ctx.stream());
-
-        return cuda_stream->interop_task([=, this](::sycl::handler &cgh) {
-            auto arg_src = CTX_IN_SYCL_MEMORY(DNNL_ARG_SRC);
-            auto arg_wt = CTX_IN_SYCL_MEMORY(DNNL_ARG_WEIGHTS);
-            auto arg_dst = CTX_OUT_SYCL_MEMORY(DNNL_ARG_DST);
 
-            auto arg_bias = xpu::sycl::interop_memory_arg_t<
-                    ::sycl::access::mode::read>();
-            auto arg_scratch = xpu::sycl::interop_memory_arg_t<
-                    ::sycl::access::mode::read_write>();
+    ~cudnn_matmul_lt_runtime_args_exec_t() = default;
 
-            auto arg_src_scale
-                    = CTX_IN_SYCL_MEMORY(DNNL_ARG_ATTR_SCALES | DNNL_ARG_SRC);
-            auto arg_wei_scale = CTX_IN_SYCL_MEMORY(
-                    DNNL_ARG_ATTR_SCALES | DNNL_ARG_WEIGHTS);
-            auto arg_dst_scale
-                    = CTX_IN_SYCL_MEMORY(DNNL_ARG_ATTR_SCALES | DNNL_ARG_DST);
+protected:
+    void init_scratch_buffers(std::size_t reorder_scratch_size,
```

Comment:
I have one concern with the way how we work with the buffers here. They are stored in the primitive object that could be used from multiple threads simultaneously. What is going to happen if we execute the primitive with runtime dimensions from multiple threads? It seems that it could lead to a race condition and result in an undefined behavior.

In general, all primitives in oneDNN are immutable meaning that their state cannot be changed after the initialization is done. This is required for thread safety. The implemented approach breaks the rule.
 
From SYCL specification: 4.7.2.3. Buffer synchronization rules
> The basic rule for the blocking behavior of a buffer destructor is that it blocks if there is some data to write back because a write accessor on it has been created

So, if we move the buffers from the primitive to the execute function then the execution function will become blocking because the buffers will have to be destroyed upon its exit, which doesn't seem like a good solution to me.

I can only see one option to address the issue. We could allocate buffers in the execution function but do not destroy them upon return. Instead, we could destroy them from the host task task, right after the cublaslt operation is finished (we would have to use `cuStreamSynchronize` for that).

Any thoughts?

Reply:
The new approach is to use local usm pointers which are allocated and released in the interop task, this should ensure the memory required for the specific `execute` call is native to the thread calling it.

Reply:
Looks good to me, thanks.